File: algorithm_from_2021/src/main/kotlin/self_boj/2470_두_용액.kt
package self_boj

import kotlin.math.abs

private var n = 0
private lateinit var arr: Array<Solution>

private data class Solution(val spec: Int, val isAcid: Boolean)

private fun main() {
    n = readLine()!!.toInt()
    arr = readLine()!!.split(" ")
            .map {
                val isAcid = !it.startsWith('-')
                Solution(abs(it.toInt()), isAcid)
            }
            .sortedBy { it.spec }
            .toTypedArray()

    var ans = Pair(0, Int.MAX_VALUE)
    for(i in 0 until arr.size - 1) {
        val (sol1, sol2) = parseSolution(arr[i]) to parseSolution(arr[i + 1])
        val sum = sol1 + sol2

        if(abs(ans.first + ans.second) >= abs(sum)) {
            ans = Pair(sol1, sol2)
        }
    }

    if(ans.first < ans.second) {
        print("${ans.first} ${ans.second}")
    } else {
        print("${ans.second} ${ans.first}")
    }

}

private fun parseSolution(sol: Solution): Int {
    return if(sol.isAcid) sol.spec
    else sol.spec * -1
}

File: js-boj/src/2470_두_용액.js
const rl = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
})

rl.on('line', (input) => {
  if(!n) n = +input;
  else {
    arr = input.split(' ').map(i => +i);
    rl.close();
  }
}).on('close', () => {
  console.log(sol().join(' '));
})

let n;
let arr;

const sol = () => {
  arr.sort((a, b) => a - b);
  let [start, end] = [0, n - 1];
  let near = 2000000001;
  let ans = [];

  while(start !== end) {
    const sum = arr[start] + arr[end];
    if(Math.abs(sum) < near) {
      near = Math.abs(sum);
      ans = [arr[start], arr[end]];
    }

    if(sum === 0) {
      return [arr[start], arr[end]];
    }
    else if(sum > 0) {
      end--;
    }
    else {
      start++;
    }
  }

  return ans;
}